### ISSUE TITLE
[in_app_purchase]: `ProrationMode.immediateWithTimeProration` not working for upgrading/downgrading a subscription

### ISSUE THREAD

**Issue by rohansohonee** [flutter/flutter#88802]
When using the `ProrationMode.immediateWithTimeProration` for making subscription upgrade/downgrade purchases, I am always facing the issue:
- Error, something went wrong or the Error DF- DFERH-01.

But when using `ProrationMode.immediateAndChargeProratedPrice` mode then it works correctly only when the new subscription price per time unit is greater than the previous subscription price per time unit.

I have already read through the docs: https://developer.android.com/google/play/billing/subscriptions, and from there it was understood by me how the `ProrationMode.immediateAndChargeProratedPrice` is working. But now it is unclear to me how the remaining modes are working.

Eg of my subscription plans:
1. plan1: ₹137.75 (monthly)
2. plan2: ₹207.00 (monthly)

- I purchased the plan1 first. Here I have not used any proration mode.
- Next I perform the operation to purchase plan2. So this will be the upgrade/downgrade portion of the subscription.
-  I constructed the `GooglePlayPurchaseParam` and provided the `ChangeSubscriptionParam` with the previous purchase details i.e plan1 and the proration mode `ProrationMode.immediateWithTimeProration`(did not work) or `ProrationMode.immediateWithoutProration`(did not work). 

Can it please be explained how these modes work and provide further detailed information about the errors.

>
```
Flutter 2.5.0-6.0.pre • channel dev 

in_app_purchase: 1.0.6
```

**Comment by TahaTesser** [comment]
Hi @rohansohonee 
Thanks for filing the issue, 


> When using the ProrationMode.immediateWithTimeProration for making subscription upgrade/downgrade purchases, I am always facing the issue:
> 
> Error, something went wrong or the Error DF- DFERH-01.

can you please provide a, complete reproducible minimal code sample to reproduce this?
Thank you

**Comment by rohansohonee** [comment]
I have followed the codelabs: https://codelabs.developers.google.com/codelabs/flutter-in-app-purchases#0
- The project can be cloned:
`git clone https://github.com/flutter/codelabs.git flutter-codelabs`
- Use the folder "complete"
- Then modify only the "buy" method under "lib/logic/dash_purchases.dart"

I have added the upgrade/downgrade code for the subscriptions inside the `buy` method which is present in `lib/logic/dash_purchases.dart` file.

```dart
Future<void> buy(PurchasableProduct product) async {
  // changed this from final to var
  var purchaseParam = PurchaseParam(productDetails: product.productDetails);
  switch (product.id) {
      case storeKeyConsumable:
        await iapConnection.buyConsumable(purchaseParam: purchaseParam);
        break;
      case storeKeySubscription:
      case storeKeyUpgrade:
        // added new upgrade/downgrade code only for the subscriptions
        GooglePlayPurchaseDetails? oldPurchaseDetails;

        var result = await iapConnection.getPlatformAddition<InAppPurchaseAndroidPlatformAddition>().queryPastPurchases();

         if (result.pastPurchases.isNotEmpty) {
          // select the first item from the list for this test
          oldPurchaseDetails = result.pastPurchases.first;
         }

         if (oldPurchaseDetails != null) {
          purchaseParam = GooglePlayPurchaseParam(
           productDetails: product.productDetails,
           changeSubscriptionParam: ChangeSubscriptionParam(
             oldPurchaseDetails: oldPurchaseDetails,
             prorationMode: ProrationMode.immediateWithTimeProration,
           ),
          );
         }
        }
        // end of upgrade/downgrade code
        await iapConnection.buyNonConsumable(purchaseParam: purchaseParam);
        break;
      default:
        throw ArgumentError.value(
            product.productDetails, '${product.id} is not a known product');
    }
}
```

**Comment by rohansohonee** [comment]
Hi @TahaTesser 

I have tested my subscription plans by creating a native android project (non flutter project). I have used the latest billing library version "4.0.0"

I have reproduced the same issue. Only "IMMEDIATE_AND_CHARGE_PRORATED_PRICE" & "IMMEDIATE_AND_CHARGE_FULL_PRICE" mode is working for me.

The billing response code when upgrading/downgrading subscription with the other proration modes results as 5 which is described in the documentation as "DEVELOPER_ERROR".

This concludes that the package does not have the issue.

Any idea why "IMMEDIATE_AND_CHARGE_PRORATED_PRICE" & "IMMEDIATE_AND_CHARGE_FULL_PRICE" mode works but the other modes don't?
